components/collator/src/comparison.rs

```diff
@@ -1543,4 +1631,808 @@ impl CollatorBorrowed<'_> {
+                        quaternaries,
+                        q,
+                        common_quaternaries,
+                        QUAT_COMMON,
+                        QUAT_COMMON[WEIGHT_LOW]
+                    );
+                    quaternaries.append_byte(q);
+                }
+            }
+
+            if (non_primary.bits() >> 24) as u8 == LEVEL_SEPARATOR_BYTE {
+                break; // ce == NO_CE
+            }
+        }
+
+        macro_rules! write_level {
+            ($key:ident, $level:expr, $flag:ident) => {
+                if levels & $flag != 0 {
+                    sink.write(&[LEVEL_SEPARATOR_BYTE])?;
+                    sink.write($key.as_ref())?;
+                }
+            };
+        }
+
+        write_level!(secondaries, Level::Secondary, SECONDARY_LEVEL_FLAG);
+
+        if levels & CASE_LEVEL_FLAG != 0 {
+            sink.write(&[LEVEL_SEPARATOR_BYTE])?;
+
+            // Write pairs of nibbles as bytes, except separator bytes as themselves.
+            let mut b = 0;
+            for c in cases.as_ref() {
+                debug_assert_eq!(*c & 0xf, 0);
+                debug_assert_ne!(*c, 0);
+                if b == 0 {
+                    b = *c;
+                } else {
+                    sink.write_byte(b | (*c >> 4))?;
+                    b = 0;
+                }
+            }
+            if b != 0 {
+                sink.write_byte(b)?;
+            }
+        }
+
+        write_level!(tertiaries, Level::Tertiary, TERTIARY_LEVEL_FLAG);
+        write_level!(quaternaries, Level::Quaternary, QUATERNARY_LEVEL_FLAG);
+
+        Ok(())
+    }
+}
+
+/// A [`std::io::Write`]-like trait for writing to a buffer-like object.
+///
+/// (This crate does not have access to [`std`].)
+pub trait CollationKeySink {
+    /// Writes a buffer into the writer.
+    fn write(&mut self, buf: &[u8]) -> Result<(), core::fmt::Error>;
+}
+
+impl CollationKeySink for Vec<u8> {
+    fn write(&mut self, buf: &[u8]) -> Result<(), core::fmt::Error> {
+        self.extend_from_slice(buf);
+        Ok(())
+    }
+}
+
+impl<const N: usize> CollationKeySink for SmallVec<[u8; N]> {
+    fn write(&mut self, buf: &[u8]) -> Result<(), core::fmt::Error> {
+        self.extend_from_slice(buf);
+        Ok(())
+    }
+}
+
+trait CollationKeySinkExt {
+    /// Write a single byte into the writer.
+    fn write_byte(&mut self, b: u8) -> Result<(), core::fmt::Error>;
+
+    /// Write leading bytes up to a zero byte, but always write at least one byte.
+    fn write_to_zero(&mut self, buf: &[u8]) -> Result<(), core::fmt::Error>;
+}
+
+impl<T> CollationKeySinkExt for T
+where
+    T: CollationKeySink,
+{
+    fn write_byte(&mut self, b: u8) -> Result<(), core::fmt::Error> {
+        self.write(&[b])?;
+        Ok(())
+    }
+
+    // guaranteed to get at least one byte and the rest are checked incrementally
+    #[allow(clippy::indexing_slicing)]
+    fn write_to_zero(&mut self, buf: &[u8]) -> Result<(), core::fmt::Error> {
```

Comment:
Please inline this to the single (non-test) call site, so that this can be removed from the trait.